File: SDE_sheet/Linked-List/12_flatten/solution.js
// User function Template for javascript

/*LINKED LIST NODE
class Node {
  constructor(x){
    this.data = x;
    this.next = null;
    this.bottom = null;
  }
}
*/

/**
 * @param {Node} head
 * @return {Node}
 */


class Solution {
    mergeTwoLists(a, b) {
        let temp = new Node(0);
        let res = temp;
        while (a !== null && b !== null) {
            if (a.data < b.data) {
                temp.bottom = a;
                temp = temp.bottom;
                a = a.bottom;
            } else {
                temp.bottom = b;
                temp = temp.bottom;
                b = b.bottom;
            }
        }
        if (a !== null) {
            temp.bottom = a;
        } else {
            temp.bottom = b;
        }
        return res.bottom;
    }

    flatten(root) {
        if (root === null || root.next === null) {
            return root;
        }
        root.next = this.flatten(root.next);
        root = this.mergeTwoLists(root, root.next);
        return root;
    }
}

File: SDE_sheet/Linked-List/12_flatten/solution.cpp
struct Node{
    int data;
    struct Node * next;
    struct Node * bottom;

    Node(int x){
        data = x;
        next = NULL;
        bottom = NULL;
    }

};
*/

class Solution {
public:
    Node* mergeTwoLists(Node* a, Node* b) {
        Node* temp = new Node(0);
        Node* res = temp;

        while (a != nullptr && b != nullptr) {
            if (a->data < b->data) {
                temp->bottom = a;
                temp = temp->bottom;
                a = a->bottom;
            } else {
                temp->bottom = b;
                temp = temp->bottom;
                b = b->bottom;
            }
        }

        if (a != nullptr) {
            temp->bottom = a;
        } else {
            temp->bottom = b;
        }

        return res->bottom;
    }

    Node* flatten(Node* root) {
        if (root == nullptr || root->next == nullptr) {
            return root;
        }

        root->next = flatten(root->next);
        root = mergeTwoLists(root, root->next);

        return root;
    }
};
